### PR TITLE
Add docs for data migrations

### PR DESCRIPTION
**Pull request by ziima**
### Description
I added a small docs regarding data migrations, based on #972.

### Checklist

This pull request is:

- [X] A documentation / typographical error fix
	- Good to go, no issue or tests are needed

**Have a nice day!**

### CODE REVIEW

File: docs/build/cookbook.rst [hunk 1]
```diff
@@ -1573,4 +1573,42 @@ the same ``env.py`` file can be invoked using asyncio as::
...
-    asyncio.run(run_async_upgrade())
\ No newline at end of file
+    asyncio.run(run_async_upgrade())
+
+
+Data migrations
+===============
+
+Alembic migrations are designed for schema migrations.
+The nature of data migrations are inherently different and it's not in fact possible in the general case to write data migrations that integrate with Alembic's schema versioning model.
```
Comment (CaselIT) [inline docs/build/cookbook.rst:1583]: let's be less definitive here:

> it's not in fact possible in the general case

it's not in fact advisable in the general case

File: docs/build/cookbook.rst [hunk 2]
```diff
@@ -1573,4 +1573,42 @@ the same ``env.py`` file can be invoked using asyncio as::
...
+For example downgrades are difficult to address since they might require deletion of data, which may even not be possible to detect.
+
+.. warning::
+
+  The solution needs to be designed specifically for each individual application and migration.
+  There are no general rules and following text is only a recommendation based on experience.
```
Comment (CaselIT) [inline docs/build/cookbook.rst:1589]:
```suggestion
  There are no general rules and the following text is only a recommendation based on experience.
```

File: docs/build/cookbook.rst [hunk 3]
```diff
@@ -1573,4 +1573,42 @@ the same ``env.py`` file can be invoked using asyncio as::
...
+3. Run the final alembic migrations (database constraints, delete columns etc.)
+
+The data migration script may also need a separate ORM model to handle intermediate state of the database.
+
+Online migration
+----------------
+The application maintains a version of schema with both versions.
+Writes are performed on both places, while the background script move all the remaining data across.
+This technique is very challenging and time demanding.
+However, it may be facilitated using `hybrid attributes <https://docs.sqlalchemy.org/orm/extensions/hybrid.html>`_.
```
Comment (CaselIT) [inline docs/build/cookbook.rst:1614]: not sure what's the connection with hybrid attributes here. not saying that it needs to be removed, but maybe the connection should be spelled out?

**Review by CaselIT — COMMENTED** [review]
Thanks for the effort, I've left a few comments

File: docs/build/cookbook.rst (hunk 2 above)
Comment (ziima) [reply docs/build/cookbook.rst]: Fixed.

File: docs/build/cookbook.rst (hunk 3 above)
Comment (ziima) [reply docs/build/cookbook.rst]: There was probably some non-trivial idea behind, which I already forgot. I just deleted it.

**Review by CaselIT — APPROVED** [review]
ok for me.

@zzzeek can you take a look?

**Comment by zzzeek** [comment]
these are good, I would want to edit them stylistically so let me pull them into gerrit and I'll try to look in the coming days.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision c985e0d8096e750df35aa07dde71f432a57656b6 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change c985e0d8096e750df35aa07dde71f432a57656b6: https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck


i tried to fix jenkins but did it wrong, attempt # 2

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

File: docs/build/cookbook.rst [hunk 4]
```diff
@@ -1573,4 +1573,41 @@ the same ``env.py`` file can be invoked using asyncio as::
...
+    asyncio.run(run_async_upgrade())
+
+
+Data migrations
+===============
+
+Alembic migrations are designed for schema migrations.
+The nature of data migrations are inherently different and it's not in fact advisable in the general case to write data migrations that integrate with Alembic's schema versioning model.
+For example downgrades are difficult to address since they might require deletion of data, which may even not be possible to detect.
+
```
Comment (sqla-tester) [inline docs/build/cookbook.rst:1585]: **Federico Caselli** (CaselIT) wrote:

Ack

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

File: docs/build/cookbook.rst [hunk 5]
```diff
@@ -1573,4 +1573,41 @@ the same ``env.py`` file can be invoked using asyncio as::
...
+Small data
+----------
+Small data migrations are easy to perform, especially in cases of initial data to a new table.
+These can be handled using :meth:`.Operations.bulk_insert`.
+
+Separate migration script
+-------------------------
+One possibility is a completely separate script aside of alembic migrations.
+The complete migration is then processed in following steps:
```
Comment (sqla-tester) [inline docs/build/cookbook.rst:1601]: **Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

I've tried reviving this change.

Can you take a second look mike?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

* docs/build/cookbook.rst (line 1620): not sure what to change here to avoid the cliff

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

it's great.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895

* docs/build/cookbook.rst (line 1620): Done

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/3895 has been **merged**. Congratulations! :)

**Pull request closed without merging**